Ensure iframe requests include Referer when location.replace or location.assign is called

sideshowbarker opened the pull request:
#### 1350b5914d017fc41f92edc00eb2b2625136dcf5
<pre>
Ensure iframe requests include Referer when location.replace or location.assign is called
<a href="https://bugs.webkit.org/show_bug.cgi?id=263072">https://bugs.webkit.org/show_bug.cgi?id=263072</a>

Reviewed by Chris Dumez and Alex Christensen.

When the loader for a Frame (e.g, an iframe) has no referrer, then this change
causes the loader&apos;s referrer to be set to the URL of its parent Frame’s Document.

That in turn ensures the associated request is sent with a Referer header —
including in the case where a document calls location.replace or
location.assign on an iframe element — which makes the WebKit behavior
in this case interoperable with existing behavior in Gecko and Blink.

Otherwise, without this change, no Referer header is sent in the
associated request, which breaks interoperability with Gecko and Blink.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-top-to-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-top-to-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-with-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-with-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/replace-with-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/replace-with-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-contents.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-postmessage-to-parent-parent.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-with-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/replace-or-assign-call-on-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/no_window_open_when_term_nesting_level_nonzero.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-expected.txt:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::setLocation):

Canonical link: <a href="https://commits.webkit.org/270741@main">https://commits.webkit.org/270741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d41cba7dc9c0d13b7b01e2b01e03c51cbac1ce8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26076 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27350 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28172 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23877 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26398 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6451 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2104 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23939 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28748 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23404 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29460 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23785 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3202 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1403 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23152 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6314 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3664 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3523 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/Location.cpp:
```diff
@@ -255,6 +255,10 @@ ExceptionOr<void> Location::replace(LocalDOMWindow& activeWindow, LocalDOMWindow
+    // If the loader for the active window's browsing context (Frame) has no outgoing referrer, but has a parent
+    // navigable, then set the loader's outgoing referrer to the URL of the parent navigable's active document.
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty() && frame->tree().parent())
+        activeWindow.frame()->loader().setOutgoingReferrer(dynamicDowncast<LocalFrame>(frame->tree().parent())->document()->urlForBindings());
```
This seems wrong to me. What if you have three documents, a parent with two children, and one of the children navigates the other. Surely the referrer shouldn't be the parent?

https://html.spec.whatwg.org/#location-object-navigate looks at the incumbent, which I'm not sure is implemented in WebKit, but it's definitely not the parent. Not sure if there's something that approximates it.

I guess ideally in our code `setLocation()` would take care of it, similar to how it is in HTML. So all the individual callers don't have to try to account for it.

sideshowbarker replied:
Thanks — I’ve now:

- updated the code to not get the URL from the parent, but instead from the _firstWindow_
- changed the explanation in the code comment and in the commit message
- fixed the test, to use `async_test` and `t.step_func` and `t.done`

annevk commented on Source/WebCore/page/Location.cpp:
```diff
@@ -255,6 +255,11 @@ ExceptionOr<void> Location::replace(LocalDOMWindow& activeWindow, LocalDOMWindow
+    // outgoing referrer to the URL of firstWindow's frame's (browsing context's) Document — which is the
+    // incumbent global object's associated Document. https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(firstFrame->document()->urlForBindings());
```
It seems this doesn't take referrer policy into account. Wouldn't that result in problems?

annevk replied:
Hmm, I haven't looked, but maybe that happens later and this is fine therefore. Not sure.

sideshowbarker replied:
As far as referrer-policy handling, it seems things go something like this:
```
Location::replace
  LocalDOMWindow::setLocation
    NavigationScheduler::scheduleLocationChange
      FrameLoader::changeLocation
        FrameLoader::loadFrameRequest
```
…and in `FrameLoader::loadFrameRequest`, here:

https://github.com/WebKit/WebKit/blob/8a62a65095faf477ebaeb47b9d236b92a4d910d1/Source/WebCore/loader/FrameLoader.cpp#L1317-L1320

…is where any referrer policy gets evaluated, and the `Referer` header is actually set (or not), based on the policy.

annevk commented on Source/WebCore/page/Location.cpp:
```diff
@@ -255,6 +255,11 @@ ExceptionOr<void> Location::replace(LocalDOMWindow& activeWindow, LocalDOMWindow
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(firstFrame->document()->urlForBindings());
     // We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame.
     frame->window()->setLocation(activeWindow, completedURL, SetLocationLocking::LockHistoryAndBackForwardList);
```
So why don't we do the same thing as `assign()` and pass _firstWindow_ as second argument here? That way `setLocation()` would take care of it, no?

sideshowbarker replied:
> So why don't we do the same thing as `assign()` and pass _firstWindow_ as second argument here? That way `setLocation()` would take care of it, no?

After looking at the code a bit, the short answer is: if we wanted to make it possible for `Location::replace` to pass a _firstWindow_ argument along — the way `Location::assign` does — then it looks like we’d basically need to either:

- change the parameter list in the existing `Location::setLocation` declaration to add a `SetLocationLocking = SetLocationLocking::LockHistoryBasedOnGestureState` default argument.

- add an additional, new `Location::setLocation` declaration like this:
  ```
  setLocation(LocalDOMWindow& incumbentWindow, LocalDOMWindow& firstWindow, const String&,
      SetLocationLocking locking);
  ```
  …that is, a function that takes _both_ a _firstWindow_ parameter and a _locking_ parameter.

…and then change `Location::replace` to call that new/modified `Location::setLocation` — rather than directly calling `DOMWindow::setLocation` as the existing `Location::replace` does now.

But that would also invalidate the code comment at https://github.com/WebKit/WebKit/blob/fcf3a66b27a8ee289b8366b7fa13689d2acf6394/Source/WebCore/page/Location.cpp#L258C22-L259:

> _We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame._

So I don’t see a way to easily make `Location::replace` do the same thing as what `Location::assign` is doing, as far as passing a  _firstWindow_ — that is, I don’t see a way to do it without it requiring a lot more changes than what the existing patch in this PR branch does, and not without invalidating that _“We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame”_ comment/requirement.

sideshowbarker replied:
…and to be clear, along with what that _“We call LocalDOMWindow::setLocation directly here because replace() always operates on the current frame”_ comment says, it seems like the more-specific reason that `Location::replace` can’t do the same thing as what `Location::assign` is doing is:

`Location::replace` needs to make `DOMWindow::setLocation` get called with `SetLocationLocking::LockHistoryAndBackForwardList` rather than with the default `SetLocationLocking::LockHistoryBasedOnGestureState` that `Location::assign` and all the other related `Location` member functions end up making `DOMWindow::setLocation` get called with.

annevk replied:
I'll defer to @cdumez, but from an HTML Standard perspective `assign()` and `replace()` call into the same underlying algorithm in the same way, with `replace()` indeed having a different history value (and no security check).

sideshowbarker replied:
> from an HTML Standard perspective `assign()` and `replace()` call into the same underlying algorithm in the same way

OK, given that — and also given that testing reveals `location.assign` has the same interop problem as `location.replace`: no `Referer` header when a document calls it from an iframe — I’ve updated the patch to move the change out of the `Location::replace` code and into the `LocalDOMWindow::setLocation` code. So that causes the same (code) algorithm to now get applied for both `location.replace` and `location.assign` handling.

I also updated the WPT test to include a `location.assign` case.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, set its outgoing referrer
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
```
`if (RefPtr activeFrame = activeWindow.frame(); activeFrame && activeFrame->loader().outgoingReferrer().isEmpty())` to be safe.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(dynamicDowncast<LocalFrame>(page()->mainFrame())->document()->urlForBindings());
```
`activeFrame->checkedLoader().setOutgoingReferrer`

cdumez replied:
Also, you can get the main frame like so: `frame->mainFrame()` so you don't have to null check the `page()` too.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(dynamicDowncast<LocalFrame>(page()->mainFrame())->document()->urlForBindings());
```
You should probably null check `dynamicDowncast<LocalFrame>(page()->mainFrame())` too.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,13 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
     if (completedURL.protocolIsJavaScript() && frameElement() && !frameElement()->document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, completedURL.string(), frameElement()))
         return;
 
+    RefPtr mainFrame = dynamicDowncast<LocalFrame>(frame->mainFrame());
```
I would think we want the parent frame rather than the main frame.  If a.com has an iframe with b.com that has an iframe with c.com, then a.com didn't "refer" to c.com.  b.com did, right?

achristensen07 replied:
Anne seems to disagree.  He's probably right, but I'm not sure.

annevk replied:
Your scenario is for the initial navigation which is based on the attribute value. That's different. In this case it's really about who called the method and there's no reason for it to be different from the other `Location` methods. (Which in part is why I argued for refactoring it so they all invoke the same primitive.)

sideshowbarker replied:
OK, further testing indicates that the parent frame is actually what Gecko and Blink set the Referer from.

So, because the purpose of this change has, so far, been to achieve interoperability/compat with Gecko and Blink, I’ve updated the implementation to use the parent frame rather than the “main” frame — and I’ve added more tests.

> If a.com has an iframe with b.com that has an iframe with c.com, then a.com didn't "refer" to c.com. b.com did, right?

Right it seems — yes, as far as what happens when script running in b.com calls `location.replace` or  `location.assign` are called on a c.com iframe.

Specifically:
1. The `assign-with-nested-iframe.html` and `replace-with-nested-iframe.html` tests I added are the a.com top-level documents that have a b.com iframe (`resources/replace-or-assign-call-on-iframe.html`) that in turn also has an iframe (c.com).
2. Script running in the b.com (`resources/replace-or-assign-call-on-iframe.html`) iframe calls  `location.replace` or  `location.assign` on the c.com iframe
3. The Referer for the c.com request in Gecko and Blink is the parent b.com (`resources/replace-or-assign-call-on-iframe.html`) iframe — _not_ the top-level `assign-with-nested-iframe.html` and `replace-with-nested-iframe.html` documents (“main” frames).

It’s not clear to me which part of the spec states what the required behavior is for that case — and not clear to me if Gecko and Blink behavior is violating the relevant spec requirements.

But if we want to have interop/compat with Gecko and Blink on this, then it seems we need to use the parent frame as the referrer in the above case. And from the spec side, if it seems desirable to have the spec match the existing Gecko and Blink behavior, and the spec needs to change to do that — then I’d be happy to also work on spec patch.

If on the other hand, the existing Gecko and Blink behavior here seems undesirable — if it seems we don’t want WebKit to match that — then I could change the implementation here back to using whatever the current spec actually requires (which is maybe the “main” window, but I’m still actually not clear if that’s what the spec says…).

And in that case, I guess I’d also need to raise Gecko and Blink bugs to try to get them to change (again, with the goal still being to get interop on this).

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,13 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (RefPtr activeFrame = activeWindow.frame(); activeFrame && activeFrame->loader().outgoingReferrer().isEmpty() && mainFrame)
+        activeFrame->loader().setOutgoingReferrer(mainFrame->document()->urlForBindings());
```
urlForBindings looks a little strange here.  This is not the bindings.  We're also missing a call to strippedForUseAsReferrer.  Also, I think we may want to restrict this to just the origin, rather than sending the whole path and query as a referrer.

Could you also add tests for the cases I mentioned?

sideshowbarker replied:
> urlForBindings looks a little strange here. This is not the bindings.

Thanks, yeah, changed to just `url()`

> We're also missing a call to strippedForUseAsReferrer

OK, added

> Also, I think we may want to restrict this to just the origin, rather than sending the whole path and query as a referrer.

Gecko and Blink send the whole path and query — so in order to have interop/compat with their existing behavior, the change here matches that.

But if we decide it’s desirable to not send the whole path and query, then I could switch the implementation to doing that — and then, I guess, raise Gecko and Blink bugs to try to get them to change (again, with the goal still being to get interop on this).

> Could you also add tests for the cases I mentioned?

See my previous comment — I added more tests (and the added tests show that Gecko and Blink are using the parent frame URL to set the Referer, not the “main” frame URL).

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, set its outgoing referrer
+    // to the URL of its parent frame's Document.
+    if (RefPtr activeFrame = activeWindow.frame(); activeFrame && activeFrame->loader().outgoingReferrer().isEmpty() && parent)
+        activeFrame->loader().setOutgoingReferrer(document()->completeURL(dynamicDowncast<LocalFrame>(parent)->document()->url().strippedForUseAsReferrer()));
```
Thanks for diligently looking into this.

If the parent is not a LocalFrame, we don't want to crash.  This would happen if the referrer is empty and the iframe is cross-site and site-isolation (currently an experimental feature) is turned on.  Please change to something like this before merging:

RefPtr localParent = dynamicDowncast<LocalFrame>(frame->tree().parent());
if (... && localParent)
    ...completeURL(localParent->document()...)

sideshowbarker committed:
Ensure iframe requests include Referer when location.replace or location.assign is called
https://bugs.webkit.org/show_bug.cgi?id=263072

Reviewed by Chris Dumez and Alex Christensen.

When the loader for a Frame (e.g, an iframe) has no referrer, then this change
causes the loader's referrer to be set to the URL of its parent Frame’s Document.

That in turn ensures the associated request is sent with a Referer header —
including in the case where a document calls location.replace or
location.assign on an iframe element — which makes the WebKit behavior
in this case interoperable with existing behavior in Gecko and Blink.

Otherwise, without this change, no Referer header is sent in the
associated request, which breaks interoperability with Gecko and Blink.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-top-to-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-replace-from-top-to-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-with-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/assign-with-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/replace-with-nested-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/replace-with-nested-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-contents.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-postmessage-to-parent-parent.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/iframe-with-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-location-interface/resources/replace-or-assign-call-on-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/no_window_open_when_term_nesting_level_nonzero.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-expected.txt:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::setLocation):

Canonical link: https://commits.webkit.org/270741@main
